### ISSUE TITLE
Enforce that `interface{}` is spelled `any`

### ISSUE THREAD

**Issue by cespare** [dominikh/go-tools#1344]
I propose to flag when `interface{}` is written instead of `any`. Maybe for Go 1.20+ or so, since it was added in Go 1.18?

It's easy to fix up a codebase to use `any` with, for instance,

    gofmt -w -r 'interface{} -> any' .

I suspect this might be controversial but I wanted to throw it out there since I'm implementing this now in our company-internal linter.

(I feel like there must have been previous discussion about this, either as part of staticcheck or as part of vet or even as an automatic gofmt feature, but my searches haven't turned up anything. It's difficult to search for these terms.)

**Comment by dominikh** [comment]
This is definitely (too) controversial for a default check, it's pretty much nitpicking existing code for no practical reason. We could maybe consider it as an optional style check, but I'd like to delay even that by a couple more versions.

As for prior discussion: https://github.com/golang/go/issues/33232#issuecomment-923069224

**Comment by dominikh** [comment]
@cespare does the existence of `modernize` (`golang.org/x/tools/gopls/internal/analysis/modernize/cmd/modernize`) address this issue?

**Comment by ccoVeille** [comment]
gofmt also supports it when configured to do so

```console 
$ gofmt -w -r 'interface{} -> any' src/
```

**Comment by cespare** [comment]
@ccoVeille yes, my issue description mentions that too. This is about enforcing it as a style check.

**Comment by cespare** [comment]
@dominikh 

I'm not sure I would want to enforce "`modernize` produces no results" as a CI check at my company yet. I just tried it out and immediately ran into [this false positive](https://github.com/golang/go/issues/72917) until I switched to the tip version. And in general it seems a little aggressive to be enforcing automatically.

It's definitely an exciting tool and I'm planning to start using it, at least for one-off modernization improvements.

As far as this issue goes, as I mentioned in my original report I wrote our own internal linter for `interface{}` and that has been working fine for the past few years. I'm happy to just keep that going forward until/unless we switch to automatically-enforce `modernize`.

On my own personal projects I'd probably be fine with just running `modernize` automatically.

So if you think this isn't a good fit for a staticcheck style lint, feel free to close this.

**Comment by ccoVeille** [comment]
> @ccoVeille yes, my issue description mentions that too. This is about enforcing it as a style check.

Indeed, I'm sorry I read too fast.

I know these two exists in:

- gopls https://pkg.go.dev/golang.org/x/tools/gopls/internal/analysis/useany

- revive: https://github.com/mgechev/revive/blob/HEAD/RULES_DESCRIPTIONS.md#use-any

**Comment by dominikh** [comment]
@cespare `modernize` has a `-category` flag that restricts which changes are made/flagged, and each kind of modernization has its own category. `interface{}` -> `any` is `efaceany`, for example.

**Comment by dominikh** [comment]
The existence of `modernize` has nudged me towards not including this in Staticcheck.

**Comment by cespare** [comment]
@dominikh oh, very nice. I could definitely see adding modernize with a whitelist of categories. (That's basically what we do with vet checks.)
